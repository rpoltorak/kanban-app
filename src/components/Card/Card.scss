.card {
  position: relative;
  background-color: #fff;
  box-shadow: 0 1px 0 rgba(9,45,66,.25);
  border-radius: 3px;
  margin-top: 10px;
  padding: 6px 8px 2px;
  cursor: pointer;

  &__title {
    font-weight: bold;
    padding-bottom: 7px;
    margin-bottom: 7px;
    border-bottom: 1px solid #eaeaea;
  }
}

