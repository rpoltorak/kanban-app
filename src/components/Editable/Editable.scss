.editable {
  &__input {
    font-size: 14px;
    line-height: 1.5;
    min-height: 20px;
    max-width: 100%;
    resize: none;
    padding: 2px 5px;

    &:focus {
      outline: 0;
      border-bottom: 1px solid #000;
    }

    &:empty:before {
      content:attr(placeholder);
      color: #aaa;
    }
  }

  &__wrapper {
    margin-bottom: 5px;
  }

  &__counter {
    font-size: 12px;
    text-align: right;
  }
}

.error {
  color: red;

  .editable {
    &__input {
      border-color: red;
    }
  }
}


