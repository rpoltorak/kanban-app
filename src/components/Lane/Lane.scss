.lane {
  position: relative;
  background-color: #dfe3e6;
  width: 220px;
  min-height: 30px;
  color: #000;
  border-radius: 0 0 3px 3px;
  padding: 15px;
  margin-right: 20px;

  &__title {
    .editable {
      font-size: 16px;
      font-weight: bold;
    }
  }
}
