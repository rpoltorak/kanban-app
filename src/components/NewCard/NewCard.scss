.lane__new-card {
  margin-top: 10px;
}
