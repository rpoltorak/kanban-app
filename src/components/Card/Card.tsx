import React from "react";
import { connect } from "react-redux";
import { bindActionCreators, Dispatch, AnyAction } from "redux";

import { deleteCard, updateCard, DeleteCardAction, UpdateCardAction } from "../../actions/cardActions";
import { RootState } from "../../reducers/rootReducer";
import { getCardById } from "../../selectors/cardsSelector";
import Editable from "../Editable";

import "./Card.scss";

interface CardOwnProps {
  id: string;
  laneId: string;
}

interface CardStateProps {
  title: string;
  description: string;
}

interface CardDispatchProps {
  deleteCard: () => DeleteCardAction;
  updateTitle: (title: string) => UpdateCardAction;
  updateDescription: (description: string) => UpdateCardAction;
}

type CardProps = CardOwnProps & CardStateProps & CardDispatchProps;

export const Card: React.SFC<CardProps> = (props: CardProps) => (
  <div className="card">
    <div className="card__title">
      <Editable
        value={props.title}
        onChange={props.updateTitle}
        minLength={1}
      ></Editable>
    </div>
    <div className="card__description">
      <Editable
        value={props.description}
        onChange={props.updateDescription}
        placeholder="description"
      ></Editable>
    </div>
    <button className="btn--delete" onClick={props.deleteCard}>&times;</button>
  </div>
);

const mapStateToProps = (state: RootState, ownProps: CardOwnProps): CardStateProps => {
  const { title, description } = getCardById(state, ownProps.id);

  return { title, description };
};

const mapDispatchToProps = (dispatch: Dispatch<AnyAction>, ownProps: CardOwnProps): CardDispatchProps => (
  bindActionCreators({
    deleteCard: () => deleteCard(ownProps.id, ownProps.laneId),
    updateTitle: (title: string) => updateCard(ownProps.id, { title }),
    updateDescription: (description: string) => updateCard(ownProps.id, { description }),
  }, dispatch)
);

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Card);
