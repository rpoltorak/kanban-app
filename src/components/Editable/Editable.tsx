import React from "react";

import "./Editable.scss";

interface EditableProps {
  value: string;
  placeholder?: string;
  autoFocus?: boolean;
  minLength?: number;
  maxLength?: number;
  onChange: (value: string) => void;
}

interface EditableState {
  value: string;
  isEdited: boolean;
  isError: boolean;
}

export default class Editable extends React.PureComponent<EditableProps, EditableState> {
  public static defaultProps = {
    value: "",
  };

  public state = {
    value: this.props.value,
    isEdited: !this.props.value,
    isError: false,
  };

  public ref = React.createRef<HTMLDivElement>();

  public componentDidMount = () => {
    if (this.ref.current) {
      this.ref.current.innerHTML = this.state.value;
      this.checkErrors();
      if (this.props.autoFocus) {
        this.ref.current.focus();
      }
    }
  }

  public toggleEdit = () => this.setState((state: EditableState) => ({ isEdited: !state.isEdited }));

  public save = () => {
    this.toggleEdit();
    this.props.onChange(this.state.value);
  }

  public reset = () => this.setState((state: EditableState) => ({ value: this.props.value }));

  public handleBlur = async (event: React.FocusEvent<HTMLDivElement>) => {
    event.persist();

    await this.checkErrors();

    if (!this.ref.current) {
      return;
    }

    if (this.state.isError) {
      event.preventDefault();
      this.ref.current.focus();
    } else {
      this.props.onChange(this.ref.current.innerHTML);
      this.save();
    }
  }

  public checkErrors = () => {
    this.setState((state: EditableState) => ({
      isError: (!!this.props.minLength && state.value.length < this.props.minLength) ||
        (!!this.props.maxLength && state.value.length >= this.props.maxLength),
    }));
  }

  public handleChange = (event: React.FormEvent<HTMLDivElement>) => {
    const { innerText } = event.target as HTMLElement;

    this.setState((state: EditableState) => ({ value: innerText }), this.checkErrors);
  }

  public renderCounter = () => {
    return this.props.maxLength && this.state.isEdited && (
      <div className="editable__counter">
        {this.state.value.length}/{this.props.maxLength}
      </div>
    );
  }

  public render() {
    return (
      <div className={this.state.isError ? "editable__wrapper error" : "editable__wrapper"}>
        <div
          contentEditable={true}
          className="editable__input"
          ref={this.ref}
          placeholder={this.props.placeholder}
          defaultValue={this.state.value}
          onInput={this.handleChange}
          onBlur={this.handleBlur}
        />
        {this.renderCounter()}
      </div>
    );
  }
}
