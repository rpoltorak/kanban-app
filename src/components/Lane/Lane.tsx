import React from "react";
import { connect } from "react-redux";
import { bindActionCreators, Dispatch, AnyAction } from "redux";

import { DeleteLaneAction, deleteLane, updateLane, UpdateLaneAction } from "../../actions/laneActions";
import { RootState } from "../../reducers/rootReducer";
import { getLaneById } from "../../selectors/lanesSelector";

import NewCard from "../NewCard";
import Card from "../Card";
import Editable from "../Editable/Editable";

import "./Lane.scss";

interface LaneOwnProps {
  id: string;
}

interface LaneStateProps {
  title: string;
  cards: string[];
}

interface LaneDispatchProps {
  deleteLane: () => DeleteLaneAction;
  updateTitle: (title: string) => UpdateLaneAction;
}

type LaneProps = LaneOwnProps & LaneStateProps & LaneDispatchProps;

export const Lane: React.SFC<LaneProps> = (props: LaneProps) => (
  <div className="lane">
    <div className="lane__title">
      <Editable value={props.title} placeholder="title" onChange={props.updateTitle}></Editable>
      <button className="btn--delete" onClick={props.deleteLane}>&times;</button>
    </div>
    <div>
      {props.cards.map((cardId: string) => <Card key={cardId} id={cardId} laneId={props.id} />)}
    </div>
    <NewCard laneId={props.id} />
  </div>
);

const mapStateToProps = (state: RootState, ownProps: LaneOwnProps): LaneStateProps => {
  const { title, cards } = getLaneById(state, ownProps.id);

  return { title, cards };
};

const mapDispatchToProps = (dispatch: Dispatch<AnyAction>, ownProps: LaneOwnProps): LaneDispatchProps => (
  bindActionCreators({
    deleteLane: () => deleteLane(ownProps.id),
    updateTitle: (title: string) => updateLane(ownProps.id, { title }),
  }, dispatch)
);

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Lane);
