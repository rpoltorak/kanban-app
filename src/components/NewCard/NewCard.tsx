
import React from "react";
import { connect } from "react-redux";
import { bindActionCreators, Dispatch, AnyAction } from "redux";

import { createCard, CreateCardAction } from "../../actions/cardActions";
import Editable from "../Editable";

import "../Card/Card.scss";
import "./NewCard.scss";

interface NewCardOwnProps {
  laneId: string;
}

interface NewCardDispatchProps {
  createCard: (title: string, description: string, laneId: string) => CreateCardAction;
}

type NewCardProps = NewCardOwnProps & NewCardDispatchProps;

interface NewCardState {
  title: string;
  description: string;
  editMode: boolean;
}

const initialState = {
  title: "",
  description: "",
  editMode: false,
};

export class NewCard extends React.Component<NewCardProps, NewCardState> {
  public state = Object.assign({}, initialState);

  public toggleEditMode = () => this.setState((state: NewCardState) => ({ editMode: !state.editMode}));

  public reset = () => this.setState((state: NewCardState) => Object.assign({}, initialState));

  public updateStateFragment = (fieldName: string) => (value: string): void => (
    this.setState((state: NewCardState) => ({...state, [fieldName]: value }))
  )

  public handleCreate = () => {
    const { title, description } = this.state;

    this.reset();
    this.props.createCard(title, description, this.props.laneId);
  }

  public renderEditMode = () => (
    <>
      <div className="card">
        <Editable
          placeholder="title"
          onChange={this.updateStateFragment("title")}
          minLength={1}
          autoFocus
        />
        <Editable placeholder="description" onChange={this.updateStateFragment("description")} />
      </div>
      <div className="lane__new-card">
        <button className="btn btn--save" onFocus={this.handleCreate}>SAVE</button>
        <button className="btn btn--cancel" onClick={this.reset}>CANCEL</button>
      </div>
    </>
  )

  public renderCreateButton = () => (
    <button className="btn btn--new-card" onClick={this.toggleEditMode}>+ Add new card...</button>
  )

  public render = () => this.state.editMode ? this.renderEditMode() : this.renderCreateButton();
}

const mapDispatchToProps = (dispatch: Dispatch<AnyAction>): NewCardDispatchProps => (
  bindActionCreators({
    createCard,
  }, dispatch)
);

export default connect(null, mapDispatchToProps)(NewCard);
