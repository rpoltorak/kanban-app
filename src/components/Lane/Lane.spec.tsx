import React from "react";
import { shallow } from "enzyme";

import { Lane } from "./Lane";
import Editable from "../Editable";
import NewCard from "../NewCard";
import Card from "../Card";

const cards = ["21", "22", "23"];

describe("Lane Component", () => {
  test("to match snapshot", () => {
    const lane = shallow(
      <Lane
        id={"1"}
        title="some title"
        cards={cards}
        deleteLane={jest.fn()}
        updateTitle={jest.fn()}
      />,
    );

    expect(lane).toMatchSnapshot();
  });

  test("should render correctly when passed props", () => {
    const lane = shallow(
      <Lane
        id={"1"}
        title="some title"
        cards={cards}
        deleteLane={jest.fn()}
        updateTitle={jest.fn()}
      />,
    );

    expect(lane.find(".lane").exists()).toBe(true);
    expect(lane.find(".btn--delete").exists()).toBe(true);
    expect(lane.find(Card).length).toBe(cards.length);
    expect(lane.find(NewCard).length).toBe(1);
    expect(lane.find(Editable).length).toBe(1);
  });

  test("should call deleteLane function after click on delete button", () => {
    const spyOnDeleteLane = jest.fn();

    const lane = shallow(
      <Lane
        id={"1"}
        title="some title"
        cards={cards}
        deleteLane={spyOnDeleteLane}
        updateTitle={jest.fn()}
      />,
    );

    lane.find(".btn--delete").simulate("click");

    expect(spyOnDeleteLane).toHaveBeenCalled();
  });
});
