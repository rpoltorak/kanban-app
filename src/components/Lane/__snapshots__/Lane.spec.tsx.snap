// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lane Component to match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Unknown
    cards={
      Array [
        "21",
        "22",
        "23",
      ]
    }
    deleteLane={[MockFunction]}
    id="1"
    title="some title"
    updateTitle={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="lane__title"
        >
          <Editable
            onChange={[MockFunction]}
            placeholder="title"
            value="some title"
          />
          <button
            className="btn--delete"
            onClick={[MockFunction]}
          >
            ×
          </button>
        </div>,
        <div>
          <Connect(Component)
            id="21"
            laneId="1"
          />
          <Connect(Component)
            id="22"
            laneId="1"
          />
          <Connect(Component)
            id="23"
            laneId="1"
          />
        </div>,
        <Connect(NewCard)
          laneId="1"
        />,
      ],
      "className": "lane",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Editable
              onChange={[MockFunction]}
              placeholder="title"
              value="some title"
            />,
            <button
              className="btn--delete"
              onClick={[MockFunction]}
            >
              ×
            </button>,
          ],
          "className": "lane__title",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "onChange": [MockFunction],
              "placeholder": "title",
              "value": "some title",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "×",
              "className": "btn--delete",
              "onClick": [MockFunction],
            },
            "ref": null,
            "rendered": "×",
            "type": "button",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Connect(Component)
              id="21"
              laneId="1"
            />,
            <Connect(Component)
              id="22"
              laneId="1"
            />,
            <Connect(Component)
              id="23"
              laneId="1"
            />,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "21",
            "nodeType": "class",
            "props": Object {
              "id": "21",
              "laneId": "1",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "22",
            "nodeType": "class",
            "props": Object {
              "id": "22",
              "laneId": "1",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "23",
            "nodeType": "class",
            "props": Object {
              "id": "23",
              "laneId": "1",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "laneId": "1",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="lane__title"
          >
            <Editable
              onChange={[MockFunction]}
              placeholder="title"
              value="some title"
            />
            <button
              className="btn--delete"
              onClick={[MockFunction]}
            >
              ×
            </button>
          </div>,
          <div>
            <Connect(Component)
              id="21"
              laneId="1"
            />
            <Connect(Component)
              id="22"
              laneId="1"
            />
            <Connect(Component)
              id="23"
              laneId="1"
            />
          </div>,
          <Connect(NewCard)
            laneId="1"
          />,
        ],
        "className": "lane",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Editable
                onChange={[MockFunction]}
                placeholder="title"
                value="some title"
              />,
              <button
                className="btn--delete"
                onClick={[MockFunction]}
              >
                ×
              </button>,
            ],
            "className": "lane__title",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "onChange": [MockFunction],
                "placeholder": "title",
                "value": "some title",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "×",
                "className": "btn--delete",
                "onClick": [MockFunction],
              },
              "ref": null,
              "rendered": "×",
              "type": "button",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Connect(Component)
                id="21"
                laneId="1"
              />,
              <Connect(Component)
                id="22"
                laneId="1"
              />,
              <Connect(Component)
                id="23"
                laneId="1"
              />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "21",
              "nodeType": "class",
              "props": Object {
                "id": "21",
                "laneId": "1",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "22",
              "nodeType": "class",
              "props": Object {
                "id": "22",
                "laneId": "1",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "23",
              "nodeType": "class",
              "props": Object {
                "id": "23",
                "laneId": "1",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "laneId": "1",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
